joern-cli/frontends/c2cpg/src/main/scala/io/joern/c2cpg/astcreation/FullNameProvider.scala

```diff
@@ -0,0 +1,337 @@
+        val (uniqueName_, uniqueNameFullName_) = uniqueName("enum", name_, fullName_)
+        TypeFullNameInfo(uniqueName_, uniqueNameFullName_)
+      case n: ICPPASTNamespaceDefinition =>
+        val name_                              = n.getName.getLastName.toString
+        val fullName_                          = fullName(n)
+        val (uniqueName_, uniqueNameFullName_) = uniqueName("namespace", name_, fullName_)
+        TypeFullNameInfo(uniqueName_, uniqueNameFullName_)
+      case a: ICPPASTNamespaceAlias =>
+        val name_     = ASTStringUtil.getSimpleName(a.getAlias)
+        val fullName_ = fullName(a)
+        TypeFullNameInfo(name_, fullName_)
+      case s: IASTCompositeTypeSpecifier =>
+        val fullName_ = registerType(cleanType(fullName(s)))
+        val name_ = ASTStringUtil.getSimpleName(s.getName) match {
+          case n if n.isEmpty => lastNameOfQualifiedName(fullName_)
+          case other          => other
+        }
+        TypeFullNameInfo(name_, fullName_)
+    }
+  }
+
+  private def shortNameForIASTDeclarator(declarator: IASTDeclarator): String = {
+    if (ASTStringUtil.getSimpleName(declarator.getName).isEmpty && declarator.getNestedDeclarator != null) {
+      shortName(declarator.getNestedDeclarator)
+    } else {
+      ASTStringUtil.getSimpleName(declarator.getName)
+    }
+  }
+
+  private def shortNameForICPPASTFunctionDefinition(definition: ICPPASTFunctionDefinition): String = {
+    if (
+      ASTStringUtil.getSimpleName(definition.getDeclarator.getName).isEmpty
+      && definition.getDeclarator.getNestedDeclarator != null
+    ) {
+      shortName(definition.getDeclarator.getNestedDeclarator)
+    } else {
+      lastNameOfQualifiedName(ASTStringUtil.getSimpleName(definition.getDeclarator.getName))
+    }
+  }
+
+  private def shortNameForIASTFunctionDefinition(definition: IASTFunctionDefinition): String = {
+    if (
+      ASTStringUtil.getSimpleName(definition.getDeclarator.getName).isEmpty
+      && definition.getDeclarator.getNestedDeclarator != null
+    ) {
+      shortName(definition.getDeclarator.getNestedDeclarator)
+    } else {
+      ASTStringUtil.getSimpleName(definition.getDeclarator.getName)
+    }
+  }
+
+  private def shortNameForCPPASTIdExpression(d: CPPASTIdExpression): String = {
+    safeGetEvaluation(d) match {
+      case Some(evalBinding: EvalBinding) =>
+        evalBinding.getBinding match {
+          case f: CPPFunction if f.getDeclarations != null =>
+            f.getDeclarations.headOption.map(n => ASTStringUtil.getSimpleName(n.getName)).getOrElse(f.getName)
+          case f: CPPFunction if f.getDefinition != null => ASTStringUtil.getSimpleName(f.getDefinition.getName)
+          case other                                     => other.getName
+        }
+      case _ => lastNameOfQualifiedName(ASTStringUtil.getSimpleName(d.getName))
+    }
+  }
+
+  protected def shortName(node: IASTNode): String = {
```

Comment:
Having two separate code path for short and full names brings the risk that the short names differ from the corresponding parts in the full name. 
Is there anything that speaks against a shared code path and only at the end we select if short or full name are taken?

Reply:
One could do so but that would blow up that helper function both in code size as well as complexity. We have plenty of tests to ensure both are synced but its a good target for a future refactoring here.